From 813f6bb77564daa43d5a99fa2526651d56f91189 Mon Sep 17 00:00:00 2001
From: Igal Zeltser <izeltser@habana.ai>
Date: Tue, 01 Feb 2022 09:21:17 +0200
Subject: [PATCH] [SW-69581] gaudi3: fit: add dev specific functionality

Added device specific functions for handling fit images and loading
the content of the fit image.

Change-Id: Ib2ac1900bc3c8a82c4f6d652071b41ed17de0468
Signed-off-by: Igal Zeltser <izeltser@habana.ai>
---

diff --git a/arch/gaudi3/arc/common.c b/arch/gaudi3/arc/common.c
index bd28d92..f02a267 100644
--- a/arch/gaudi3/arc/common.c
+++ b/arch/gaudi3/arc/common.c
@@ -19,6 +19,7 @@
 #include <gaudi3/gaudi3_boot_info.h>
 
 #include <bitops.h>
+#include <string.h>
 
 #include "common.h"
 #include <preboot.h>
@@ -139,6 +140,40 @@
 	return HL_BOOT_INFO_VERSION;
 }
 
+void preboot_copy_image(u64 dst, const void *buf, int len, const char *arch)
+{
+	u32 arc_dst;
+
+	/* ARC_ADDR_MASK trancates the address to the 256MB region accesible by
+	 * Arc. MSB of 0x0020 points to the FW_MEM_REGION.
+	 */
+	arc_dst = ((u32)dst & ARC_ADDR_MASK) | BIT(29);
+
+	memcpy((void *)arc_dst, buf, len);
+}
+
+int preboot_decompress_image(u64 dst, const void *buf, int len,
+			     const char *arch, const char *compression)
+{
+	preboot_puts("Compression is not supported\n");
+	return 1;
+}
+
+void preboot_start_image(u64 startaddr, bool valid_arc_img)
+{
+	void (*image_entry)(void);
+	u32 arc1_addr = (u32)(startaddr);
+
+	/* TODO: add start for ARC0 and ARC2 */
+
+	/*
+	 * Jump to __reset of zephyr. After image_entry, control is not
+	 * expected to return here and continue.
+	 */
+	image_entry = (void *)(*(u32 *)arc1_addr);
+	image_entry();
+}
+
 #define GAUDI3_EFUSE_READ_RANGE_ALL	0xff00
 
 void device_init_efuse(void)
diff --git a/arch/gaudi3/arc/common.h b/arch/gaudi3/arc/common.h
index 5817bd4..cdf4f12 100644
--- a/arch/gaudi3/arc/common.h
+++ b/arch/gaudi3/arc/common.h
@@ -64,6 +64,8 @@
 #define ARC_RST_VALUE			0x110
 #define ARC_RST_VEC_ADDR		0x0
 
+#define ARC_ADDR_MASK			0xFFFFFFFull
+
 /* TODO: replace with true TS calibration values when available */
 #define TS_UNCALIB_A_G_COEFF		81100
 #define TS_UNCALIB_B_H_COEFF		255200
